domain/src/main/java/org/oppia/android/domain/exploration/ExplorationActiveTimeController.kt

```diff
@@ -0,0 +1,182 @@
+          .setTopicId(topicId)
+          .setLastUpdatedTimeMs(oppiaClock.getCurrentTimeMs())
+      }
+
+      if (isLastUpdatedTimestampStale(topicLearningTimeBuilder.lastUpdatedTimeMs)
+      ) {
+        topicLearningTimeBuilder.topicLearningTimeMs = sessionDuration
+      } else {
+        topicLearningTimeBuilder.topicLearningTimeMs += sessionDuration
+      }
+      topicLearningTimeBuilder.lastUpdatedTimeMs = oppiaClock.getCurrentTimeMs()
+
+      val topicLearningTime = topicLearningTimeBuilder.build()
+
+      val topicLearningTimeDatabaseBuilder = topicLearningTimeDatabase.toBuilder()
+        .putAggregateTopicLearningTime(topicId, topicLearningTime)
+      Pair(topicLearningTimeDatabaseBuilder.build(), TopicLearningTimeActionStatus.SUCCESS)
+    }
+    return dataProviders.createInMemoryDataProviderAsync(
+      RECORD_AGGREGATE_LEARNING_TIME_PROVIDER_ID
+    ) {
+      return@createInMemoryDataProviderAsync getDeferredResult(deferred)
+    }
+  }
+
+  private fun isLastUpdatedTimestampStale(lastUpdatedTimestamp: Long): Boolean {
+    val currentTime = oppiaClock.getCurrentTimeMs()
+    val differenceInMillis = currentTime - lastUpdatedTimestamp
+    return differenceInMillis > LEARNING_TIME_EXPIRATION_MILLIS
+  }
+
+  /** Returns the [TopicLearningTime] [DataProvider] for a specific topicId, per-profile basis. */
+  fun retrieveAggregateTopicLearningTimeDataProvider(
+    profileId: ProfileId,
+    topicId: String
+  ): DataProvider<TopicLearningTime> {
+    return retrieveCacheStore(profileId)
+      .transformAsync(RETRIEVE_AGGREGATE_LEARNING_TIME_PROVIDER_ID) { learningTimeDb ->
+        return@transformAsync AsyncResult.Success(
+          learningTimeDb.aggregateTopicLearningTimeMap[topicId]
+            ?: TopicLearningTime.getDefaultInstance()
+        )
+      }
+  }
+
+  private suspend fun getDeferredResult(
+    deferred: Deferred<TopicLearningTimeActionStatus>
+  ): AsyncResult<Any?> {
+    return when (deferred.await()) {
+      TopicLearningTimeActionStatus.SUCCESS -> AsyncResult.Success(null)
+    }
+  }
+
+  private fun retrieveCacheStore(
+    profileId: ProfileId
+  ): PersistentCacheStore<TopicLearningTimeDatabase> {
+    val cacheStore = if (profileId in cacheStoreMap) {
+      cacheStoreMap[profileId]!!
+    } else {
+      val cacheStore =
+        cacheStoreFactory.createPerProfile(
+          CACHE_NAME,
+          TopicLearningTimeDatabase.getDefaultInstance(),
+          profileId
+        )
+      cacheStoreMap[profileId] = cacheStore
+      cacheStore
+    }
+
+    cacheStore.primeInMemoryAndDiskCacheAsync(
+      updateMode = PersistentCacheStore.UpdateMode.UPDATE_IF_NEW_CACHE,
+      publishMode = PersistentCacheStore.PublishMode.PUBLISH_TO_IN_MEMORY_CACHE
+    ).invokeOnCompletion {
+      if (it != null) {
+        oppiaLogger.e(
+          "ExplorationActiveTimeController",
+          "Failed to prime cache ahead of data retrieval.",
+          it
+        )
+      }
+    }
+    return cacheStore
```

Comment:
```suggestion
    return cacheStoreMap.getOrPut(profileId) {
      cacheStoreFactory.createPerProfile(
        CACHE_NAME,
        TopicLearningTimeDatabase.getDefaultInstance(),
        profileId
      ).also { cacheStore ->
        cacheStore.primeInMemoryAndDiskCacheAsync(
          updateMode = PersistentCacheStore.UpdateMode.UPDATE_IF_NEW_CACHE,
          publishMode = PersistentCacheStore.PublishMode.PUBLISH_TO_IN_MEMORY_CACHE
        ).invokeOnCompletion {
          if (it != null) {
            oppiaLogger.e(
              "ExplorationActiveTimeController",
              "Failed to prime cache ahead of data retrieval.",
              it
            )
          }
        }
      }
    }
```

Shorter, more idiomatic, and avoids the potential code smell of a forced null check.

Reply:
I accepted this suggestion as is, thanks!

Reply:
This can be marked as resolved.